### ISSUE TITLE
Automatically fallback to a hot restart if phased restart fails

### ISSUE THREAD

**Issue by cjlarose** [puma/puma#2465]
**Is your feature request related to a problem? Please describe.**

There are a number of situations where a phased restart will fail (many of which are documented in the [restart docs][0]), the most common of which being upgrading puma itself. As an operator, I want puma to automatically perform a hot restart if the phased restart fails.

[0]: https://github.com/puma/puma/blob/61c6213fbab053a555815a2afa765bd069a2580b/docs/restart.md#phased-restart

For the applications that I maintain, my team has implemented a complicated start/stop script that utilizes `pumactl stats` to monitor the progress of a phased restart (by watching the `phase` of the workers). If the phased restart stops making progress, we issue a hot restart. I think puma would be more ergonomic to use if users didn't have to implement something similar to get this behavior.

**Describe the solution you'd like**

One of two options:

1. `pumactl phased-restart` should attempt a phased restart, but if workers continue to fail to boot for whatever reason, then the server would issue a hot restart automatically

2. Expand #2213. If `pumactl phased-restart --wait` existed and blocked until the phased restart completed successfully, then users would have an easier time implementing the hot restart fallback themselves. The next step would be to have `pumactl phased-restart --wait --timeout 30s`, which would fail with a non-zero exit code after 30 seconds if the phased-restart did not complete successfully. Finally, users could just do `pumactl phased-restart --wait --timeout 30s || pumactl restart` in order to manually implement the fallback behavior I'm describing. If we don't want to include `--timeout` ourselves, we could also just ask users to use the `timeout` command from coreutils. `timeout 30 pumactl phased-restart --wait || pumactl restart`.

**Comment by MSP-Greg** [comment]
Good idea.  Both solutions are based on `pumactl`.  Maybe add a startup option like `phased_restart_timeout` that would automatically switch to `restart`?

If the option isn't set, Puma behaves as it does currently.  With the option set, regardless of how the  `phased_restart` is triggered, it would switch to `restart`?

One issue is also exactly when the timeout begins?

**Comment by cjlarose** [comment]
Oh I like the `phased_restart_timeout` config option idea. Then users don't have to do anything special to get the timeout behavior when requesting a phased restart. It'd work for `pumactl phased-restart` with no additional options as well as with `SIGUSR1` directly and with `/phased-restart` from the control server.

> One issue is also exactly when the timeout begins?

I think it might make sense to just start the timeout when the request for a phased restart is received, but I'm open to other ideas.

**Comment by cjlarose** [comment]
Thinking about this a little more, a single `phased_restart_timeout` doesn't scale automatically with the number of workers. It'd be nice if users didn't have to adjust `phased_restart_timeout` based on the number of workers they have.

Another idea is to just leverage the existing `worker_boot_timeout`. If workers timeout while booting during a phased restart, maybe we'd just automatically fallback to a hot restart. There might not be a need for an additional config option.

**Comment by MSP-Greg** [comment]
> it might make sense to just start the timeout when the request for a phased restart is received

That's dependent on how many requests are waiting on processing.  Maybe a timer that works on when the `fork` code is actually called?  I haven't really looked at it.  Also the issue of 'our current app has a bug, restart now!'

**Comment by h0jeZvgoxFepBQ2C** [comment]
Hm not sure if this is not related, but I'm somehow against this approach:

Sometimes we have broken developments (typos, uncatched syntax errors) and phased restarts are not finishing successfull (since the environment can't be loaded completely due to the mentioned errors) -> but the old version stays online at least.

If we perform a complete restart, puma would be completely offline since it can't also load the code with a restart?

So it would be good to have an option to enable/disable the requested bahaviour at least?

**Comment by cjlarose** [comment]
@h0jeZvgoxFepBQ2C That's good feedback to have. It's true that if a worker is failing to boot as part of a phased restart, that a hot restart won't necessarily fix it, especially if it's a problem with the new version of the application.

It makes sense to make the behavior configurable so users can choose the behavior they want.

**Comment by briu** [comment]
Hey guys, what do you think about adding one more column to the stats for checking that booting is failed?

For example `latest_failed_phase` - number when all workers boot is failed and then we can compare `phase` number with `latest_failed_phase` number and if `phase` is smaller than `latest_failed_phase` we can understand that something went wrong with our puma instance.

Currently, it is possible to compare workers `phase` number with the master process `phase` number, but it is hard to understand what time lag is needed to restart workers
